### ISSUE TITLE
update Docker container offerings

### ISSUE THREAD

**Issue by tomkralidis** [geopython/pygeoapi#1753]
Currently, pygeoapi produces Docker images via [GitHub Actions](https://github.com/geopython/pygeoapi/blob/master/.github/workflows/containers.yml) to the following container registries:

- [Docker Hub](https://hub.docker.com/r/geopython/pygeoapi) (on release, and latest based on master branch)
- [GitHub Container Registry](https://github.com/geopython/pygeoapi/pkgs/container/pygeoapi) (latest based on master branch)

Currently, the latest image is ~1.5GB containing numerous optional dependencies.  It would be valuable to offer various images based on different deployment profiles/environments.

- [ ] determine images to produce (e.g. slim, default, full)
- [ ] add/update `Dockerfile` setups accordingly
- [ ] update GitHub Action for container publication
- [ ] update [documentation](https://docs.pygeoapi.io/en/latest/running-with-docker.html)

cc @justb4 @francbartoli @webb-ben @ricardogsilva

**Comment by kalxas** [comment]
I have been experimenting with a slim version of our Docker image. I agree it would be good to have different images for different use cases.

**Comment by mikebelanger** [comment]
I'd be interested in what you've found too @kalxas.

**Comment by cmotadev** [comment]
Hey guys, I’ve already did it with Debian Bookworm and python. With some
deep modifications. I’m testing some stuff but the results are promising

Greetings

Carlos Eduardo Mota


Em ter., 15 de abr. de 2025 às 13:24, Mike ***@***.***>
escreveu:

**Comment by cmotadev** [comment]
> Hey guys, I’ve already did it with Debian Bookworm and python. With some
> deep modifications. I’m testing some stuff but the results are promising
> 
> Greetings
> 
> Carlos Eduardo Mota
> 
> 
> Em ter., 15 de abr. de 2025 às 13:24, Mike ***@***.***>
> escreveu:
> […](#)

@tomkralidis @kalxas @mikebelanger 

PyGeoAPI`s debian based image, multi-staged, rootless and virtualenv-based (experimental)

https://github.com/nds-cprm/pygeoapi.git

**Comment by webb-ben** [comment]
> * determine images to produce (e.g. slim, default, full)

- `default` is the current Dockerfile and requirements-docker.txt to constitute the default image.

- `slim` would constitute anything needed to run core excluding the dependencies for `providers/*`, `processes/*`, and `/pubsub/*`.

- `full` would include django and starlette deps?

**Comment by mattdear** [comment]
Docker announced that their [hardened Docker images](https://www.docker.com/products/hardened-images/) are now free for everyone. So I was wondering if a hardened Pygeoapi image could be offered or if all image types default, slim, and full could be created using a hardened base image from Docker.

**Comment by tomkralidis** [comment]
@mattdear thanks for this, timely.  We discussed our Docker strategy at our last online meeting, and we will be putting forth a "pygeoapi Docker survey" to determine the safest path for this ticket based on user feedback, after our 0.23.0 release.

**Comment by doublebyte1** [comment]
Maybe this is a separate topic, but I think it would be nice to have images pinned down to particular releases or commit hashes. These default, slim and full change all the time, and although I see use cases for that, I think that approach may not work for everyone.

**Comment by mattdear** [comment]
> @mattdear thanks for this, timely.  We discussed our Docker strategy at our last online meeting, and we will be putting forth a "pygeoapi Docker survey" to determine the safest path for this ticket based on user feedback, after our 0.23.0 release.

Thanks @tomkralidis I will keep an eye out for the survey. 🙂

Also, I was thinking of attending the meeting on Friday but was unable to due to childcare. Is it open to anyone?

**Comment by tomkralidis** [comment]
> > [@mattdear](https://github.com/mattdear) thanks for this, timely.  We discussed our Docker strategy at our last online meeting, and we will be putting forth a "pygeoapi Docker survey" to determine the safest path for this ticket based on user feedback, after our 0.23.0 release.
> 
> Thanks [@tomkralidis](https://github.com/tomkralidis) I will keep an eye out for the survey. 🙂
> 
> Also, I was thinking of attending the meeting on Friday but was unable to due to childcare. Is it open to anyone?

Yes, meetings are open to anyone.  You can view our meeting register on the [wiki](https://github.com/geopython/pygeoapi/wiki/Meetings) (announcements sent on the mailing list).

**Comment by doublebyte1** [comment]
> Maybe this is a separate topic, but I think it would be nice to have images pinned down to particular releases or commit hashes. These default, slim and full change all the time, and although I see use cases for that, I think that approach may not work for everyone.

Apologies, I missed that you already have that! (👍🏽

**Comment by webb-ben** [comment]
This may be tied to the question about the source of the install for Docker images - but I wasn't sure if the types of images offered were also open for discussion. A thought I had while filling out the survey was an interest in better alignment between the `requirements*.txt` files and docker images. Instead of viewing `requirements-docker.txt` as a separate set of dependencies, the various images would correspond to the underlying requirements installed. This is more to demonstrate the idea, but

- slim -> requirements.txt
- standard -> requirements.txt + requirements-provider.txt
- starlette -> requirements.txt + requirements-provider.txt requirements-starlette.txt
- full -> requirements*.txt(?)
- etc

**Comment by webb-ben** [comment]
Were there any conclusions from the pygeoapi Docker survey? https://lists.osgeo.org/pipermail/pygeoapi/2026-February/000481.html

**Comment by tomkralidis** [comment]
The proposal is that Python packages are not derived from Ubuntu system dependencies.  We'll have the work in a PR once the next release is complete.

**Comment by jorgejesus** [comment]
First comment: Issue #2392  is a blocker of the new docker implementations
To test current docker proposals:
- Fork / branch: [https://github.com/terraops-org/pygeoapi/tree/docker-variants](https://github.com/terraops-org/pygeoapi/tree/docker-variants)
- README (rationale + build/test steps): [docker-proposals/README.md](https://github.com/terraops-org/pygeoapi/blob/docker-variants/docker-proposals/README.md)

Everything is self-contained under `docker-proposals/`; the current Dockerfile is untouched, so nothing in the repo changes until we agree on a direction.

All five build and serve. This also resolves #2221 (pure pip no gunicorn/gevent mismatch) and #2180 (geopandas in standard/full).

  To test:
```bash 
git clone -b docker-variants https://github.com/terraops-org/pygeoapi.git
cd pygeoapi
docker build -f docker-proposals/Dockerfile.slim -t pygeoapi:slim .
docker run -p 5000:80 pygeoapi:slim
 # tge same pattern for Dockerfile.alpine / .distroless / .standard / .full
```


These are my resilts

  | Variant | base | Python | size (uncompressed / compressed) |
  |---|---|---|---|
  | **slim** | `ubuntu:26.04` LTS | 3.14 | 429 MB / ~144 MB |
  | **distroless** (experimental) | `distroless/python3-debian13` | 3.13 | 272 MB / ~89 MB |
  | **standard** | `osgeo/gdal:ubuntu-small-3.12.4` | 3.12 | 1.07 GB |
  | **full** | `osgeo/gdal:ubuntu-small-3.12.4` | 3.12 | 1.06 GB |
  | alpine (experiment) | `python:alpine` | 3.12 | 312 MB / ~97 MB |

**Comment by justb4** [comment]
My comments. I support the idea of multiple Docker images. I would follow another approach. At least with more unification and common conventions. And maybe start with two variants and build, like **slim** and **full** and work from there on possibly intermediate variants. My feeling is that the above approach has drawbacks:

* variation in base images and install commands
* use of a mix of Debian and PyPi packages and custom compiles
* a GDAL Base image is handy, but may give other complexities (versions etc) and the size gain may be small
* these days the Ubuntu Docker Image already very small, 75-80 MB uncompressed
* all these different environments need their own security patching

IMHO the above approach will make maintenance quite hard. The mix of Debian, Ubuntu-GIS and PyPi packages and custom compiles has felt always awkward to me, at the time I did not know better..., unless we could rely completely on one repo like Debian. 

Our (pygeoapi's) main problem is that we need Python and non-Python (like GDAL) dependencies. Most Python environment managers like `venv`, `pyenv` and even `uv` cannot deal with non-Python dependencies, or take the "wheel-route". 
I shared this document several times, but for me was en eye-opener:
https://www.nijho.lt/post/python-environments/ and subsequently:
https://codecookcash.substack.com/p/simplifying-dependency-management

How I would approach this? BTW Appreciating the work @jorgejesus has already done!

I just spent quite some time on GeoHealthCheck cleaning. The main approach that helped me was to first standardize package management:

* from a bunch of requirements.txt files standardized to `pyproject.toml` also part of the Python standard (IMHO)
* introduced `Pixi` (thanks @francbartoli !) as Python Environment manager 
* all dependencies, inclding binary, multiple environments within a single `.pixi` directory no more system-wide "Proj" libs
* slimmed the Dockerfile with a Pixi multi-stage build, copy-ing even all binary dependencies
* `Pixi` is not even needed at runtime in Dockerfile

The main asset here is that Pixi is built on top of Conda and `uv` and uses both the PyPi and Conda Channels. Blazing fast as well. 
(Though I never tested with GDAL, but many Data Science folks use Conda and Mamba family.)

The whole bunch is now much better maintainable, but also for local testing, even linestepping within Pycharm and once even connecting to Docker Container. Version management in one place, warnings when misaligning versions.

Long story. But the whole approach has just two main files and a README, see these GeoHealthCheck files:

* [pyproject.toml](https://github.com/geopython/GeoHealthCheck/blob/master/pyproject.toml)
* [Dockerfile](https://github.com/geopython/GeoHealthCheck/blob/master/Dockerfile)
* [README](https://github.com/geopython/GeoHealthCheck/blob/master/README.md)

I would first start with Pixi (or maybe one from the Conda family), streamline the whole dep management, redesign the current Dockerfile and then split into the **Slim**, **Full** etc.

This is quite some effort, I probably spent days, also as the whole Pixi-Conda thing was new for me (never too old to learn!). But am very happy I did!

**Comment by tomkralidis** [comment]
I think the general principle is that Python packages are not derived from Ubuntu system dependencies.  The next question centres around.

@jorgejesus any details on using osgeo/gdal as the base image for standard/full images (vs. UbuntuGIS PPA'ing them)?  So this would mean similar to the slim option, adding GDAL/etc. via UbuntuGIS.

**Comment by justb4** [comment]
Just tested Pixi with GDAL (as GDAL is no dep in GHC) but added as test temporary. By default Pixi will use system GDAL (when added to `dependencies`) which has to match version, but it is also possible to get GDAL as a Conda package within the Pixi `.pixi/` env-bundle. This added around 150MB to the .pixi libs. Within Docker about 250MB uncompressed. A full GDAL I expect. The latter number is more reliable, as maybe are system-deps (GEOS?) may have been used.

To use GDAL from Conda channel: `pixi add GDAL`, then it is added at the right places in `pyproject.toml`. 

Edit: using GDAL Base image could still be a good option! I am staying away a bit from Alpine, unless using self-contained Image like PostGIS.

**Comment by francbartoli** [comment]
> My comments. I support the idea of multiple Docker images. I would follow another approach. At least with more unification and common conventions. And maybe start with two variants and build, like **slim** and **full** and work from there on possibly intermediate variants. My feeling is that the above approach has drawbacks:
> 
> * variation in base images and install commands
> * use of a mix of Debian and PyPi packages and custom compiles
> * a GDAL Base image is handy, but may give other complexities (versions etc) and the size gain may be small
> * these days the Ubuntu Docker Image already very small, 75-80 MB uncompressed
> * all these different environments need their own security patching
> 
> IMHO the above approach will make maintenance quite hard. The mix of Debian, Ubuntu-GIS and PyPi packages and custom compiles has felt always awkward to me, at the time I did not know better..., unless we could rely completely on one repo like Debian.
> 
> Our (pygeoapi's) main problem is that we need Python and non-Python (like GDAL) dependencies. Most Python environment managers like `venv`, `pyenv` and even `uv` cannot deal with non-Python dependencies, or take the "wheel-route". I shared this document several times, but for me was en eye-opener: https://www.nijho.lt/post/python-environments/ and subsequently: https://codecookcash.substack.com/p/simplifying-dependency-management
> 
> How I would approach this? BTW Appreciating the work [@jorgejesus](https://github.com/jorgejesus) has already done!
> 
> I just spent quite some time on GeoHealthCheck cleaning. The main approach that helped me was to first standardize package management:
> 
> * from a bunch of requirements.txt files standardized to `pyproject.toml` also part of the Python standard (IMHO)
> * introduced `Pixi` (thanks [@francbartoli](https://github.com/francbartoli) !) as Python Environment manager
> * all dependencies, inclding binary, multiple environments within a single `.pixi` directory no more system-wide "Proj" libs
> * slimmed the Dockerfile with a Pixi multi-stage build, copy-ing even all binary dependencies
> * `Pixi` is not even needed at runtime in Dockerfile
> 
> The main asset here is that Pixi is built on top of Conda and `uv` and uses both the PyPi and Conda Channels. Blazing fast as well. (Though I never tested with GDAL, but many Data Science folks use Conda and Mamba family.)
> 
> The whole bunch is now much better maintainable, but also for local testing, even linestepping within Pycharm and once even connecting to Docker Container. Version management in one place, warnings when misaligning versions.
> 
> Long story. But the whole approach has just two main files and a README, see these GeoHealthCheck files:
> 
> * [pyproject.toml](https://github.com/geopython/GeoHealthCheck/blob/master/pyproject.toml)
> * [Dockerfile](https://github.com/geopython/GeoHealthCheck/blob/master/Dockerfile)
> * [README](https://github.com/geopython/GeoHealthCheck/blob/master/README.md)
> 
> I would first start with Pixi (or maybe one from the Conda family), streamline the whole dep management, redesign the current Dockerfile and then split into the **Slim**, **Full** etc.
> 
> This is quite some effort, I probably spent days, also as the whole Pixi-Conda thing was new for me (never too old to learn!). But am very happy I did!

I'm totally supportive of this direction, I wrote almost all the migration to pixi of GHC with the experience done during OGC sprint in Nov 2025. I paste here an example of a minimal `pixi.toml` for GDAL binaries coming from different channels, the `dev` one is the conda channel from GDAL nightly builds. Hope it is useful to figure out the grade of freedom that pixi would open for developers and maintainers.

```toml
# pixi.toml
[workspace]
authors = ["Francesco Bartoli <xbartolone@gmail.com>"]
# conda-forge only at workspace level: it is the default channel
# for every environment that does not override it
channels = ["conda-forge"]
name = "pygeoapi"
# Multi-platform guarantee: pixi solves and locks every environment
# for all the platforms listed here, so pixi.lock is reproducible on
# Apple Silicon, Intel macOS, Linux (x86_64/ARM) and Windows alike
platforms = ["osx-arm64", "osx-64", "linux-64", "linux-aarch64", "win-64"]
version = "0.1.0"

# ---------------------------------------------------------------
# Dependencies and tasks shared by all environments
# ---------------------------------------------------------------
[dependencies]
python = ">=3.14"

[tasks]
gdal-version = "python -c \"from osgeo import gdal; print('GDAL version:', gdal.__version__)\""

# ---------------------------------------------------------------
# Feature "main": stable runtime, conda-forge only, GDAL 3.x
# ---------------------------------------------------------------
[feature.main.dependencies]
gdal = "<4"
libgdal-core = "<4"

# ---------------------------------------------------------------
# Feature "dev": development of embryonic features against GDAL
# builds coming from the gdal-master channel (nightly/master)
# ---------------------------------------------------------------
[feature.dev]
# priority = 1 > 0 (default for workspace channels), so in this
# environment gdal-master wins over conda-forge for the packages
# it publishes
channels = [{ channel = "gdal-master", priority = 1 }]

[feature.dev.dependencies]
# No upper bound: we want the development builds (e.g. 4.0.0devYYYYMMDD)
gdal = "*"
libgdal-core = "*"
ipython = ">=9.6.0,<10"

[feature.dev.tasks]
shell = "ipython"

# ---------------------------------------------------------------
# Environments
# ---------------------------------------------------------------
[environments]
default = { features = ["main"], solve-group = "main" }
dev = { features = ["dev"] } 
```
